Sources/Core/Renderer/GUI/GUIElement/GUITextInput.swift

```diff
@@ -12,21 +13,46 @@ struct GUITextInput: GUIElement {
     ).meshes(context: context)
 
     // Message
-    var message = try content.meshes(context: context)
-    message.translate(amount: [2, 2])
+    var firstMessage = try String(content.prefix(upTo: content.index(content.endIndex, offsetBy: -cursorIndex))).meshes(context: context)
+    var secondMessage = try String(content.suffix(from: content.index(content.endIndex, offsetBy: -cursorIndex))).meshes(context: context)
+    if firstMessage.size().x != 0 {
+      firstMessage.translate(amount: [2, 2])
+      secondMessage.translate(amount: [3 + firstMessage.size().x, 2])
+    } else {
+      firstMessage.translate(amount: [0, 2])
+      secondMessage.translate(amount: [2, 2])
+    }
+    // When at the end of the message, use an underscore cursor
+    if cursorIndex == 0 {
+      var cursor: [GUIElementMesh] = []
+      if Int(CFAbsoluteTimeGetCurrent() * 10/3) % 2 == 1 {
+        cursor = try "_".meshes(context: context)
+        let messageWidth = firstMessage.size().x + secondMessage.size().x
+        var xOffset = 2
+        if messageWidth != 0 {
+          xOffset = messageWidth + 4
+        }
+        cursor.translate(amount: [xOffset, 2])
+      }
+      let message = firstMessage + secondMessage
+      return background + message + cursor
+    // Otherwise, use a vertical bar cursor
+    } else {
+      let cursorWidth = 1
+      let cursorHeight = 9
 
-    // Cursor
-    var cursor: [GUIElementMesh] = []
-    if Int(CFAbsoluteTimeGetCurrent() * 10/3) % 2 == 1 {
-      cursor = try "_".meshes(context: context)
-      let messageWidth = message.size().x
+      var cursor = GUIRectangle(
+        size: [cursorWidth, cursorHeight],
+        color: [1, 1, 1, 1]
+      ).meshes(context: context)
       var xOffset = 2
-      if messageWidth != 0 {
-        xOffset = messageWidth + 4
+      if firstMessage.size().x != 0 {
+        xOffset = firstMessage.size().x + 2
       }
-      cursor.translate(amount: [xOffset, 2])
-    }
+      cursor.translate(amount: [xOffset, 1])
 
-    return background + message + cursor
+      let message = firstMessage + secondMessage
+      return background + message + cursor
```

Comment:
This should be rearranged slightly so that this doesn't need to be duplicated on line 38. Add `var cursor: [GUIElementMesh]` before the if-else, and then use it inside the if/else blocks. Then after the if/else you can simply `return background + message + cursor` once.

```swift
var cursor: [GUIElementMesh] = []
if Int(CFAbsoluteTimeGetCurrent() * 10/3) % 2 == 1 {
  if ... {
    cursor = ...
    cursor.translate(...)
  } else {
    cursor = ...
    cursor.translate(...)
  }
}
return background + message + cursor
```